### PR TITLE
parser, resolver: Implement Replace operator :=

### PR DESCRIPTION
**Pull request by qinqon**
For the main scenario NMPolicy need to resolve `capture.base-iface-routes | routes.running.next-hop-interface:="br1"` part of this expression is the replace operator `:=` this PR implement that operator at parser and resolver and also allow filter/replace to use a capture ref path as input so next PR implementing the pipe will involve only the parser.

Thech debt:
- For apply functions at `nmpolicy/internal/path.go`:
  - Use a struct so func and flags does not need to be passed around
  - The expectedNode argument is not needed to be passed if we function to apply is a closure containing it.

### CODE REVIEW

File: nmpolicy/internal/resolver/resolver.go [hunk 1]
```diff
@@ -93,35 +93,53 @@ func (r *resolver) resolveCaptureEntryName(captureEntryName string) (types.NMSta
...
+	if err != nil {
+		return nil, err
+	}
 
 	path, err := r.resolvePath((*operator)[1])
 	if err != nil {
-		return nil, wrapWithEqFilterError(err)
+		return nil, err
 	}
 	filteredValue, err := r.resolveFilteredValue((*operator)[2])
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/resolver.go:129]: Now it is used for replace as well, the method name should be changed.

File: nmpolicy/internal/resolver/resolver_test.go [hunk 2]
```diff
@@ -708,3 +709,47 @@ base-iface-routes:
...
+`,
+
+			expectedCapturedStates: `
+bridge-routes:
+  state:
+    routes:
+      running:
+      - destination: 0.0.0.0/0
+        next-hop-address: 192.168.100.1
+        next-hop-interface: br1
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/resolver_test.go:742]: Please add an entry with `next-hop-interface` different than `br1`. I believe the current code has a bug a this entry will be filtered.

**Review by AlonaKaplan — COMMENTED** [review]
Reviewed the first two commits.

File: nmpolicy/internal/parser/parser.go [hunk 3]
```diff
@@ -204,3 +208,31 @@ func (p *parser) parseEqFilter() error {
...
+	operator.Replace[1] = *p.lastNode
+
+	p.nextToken()
+
+	if p.currentToken().Type == lexer.STRING {
+		if err := p.parseString(); err != nil {
+			return err
+		}
+		operator.Replace[2] = *p.lastNode
+	} else if p.currentToken().Type != lexer.EOF {
```
[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/parser/parser.go:233]: Is not, we compare it with `!=`

File: nmpolicy/internal/resolver/replace.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package resolver
+
+import (
+	"github.com/nmstate/nmpolicy/nmpolicy/internal/ast"
+)
+
+func replace(inputState map[string]interface{}, path ast.VariadicOperator, expectedNode ast.Node) (map[string]interface{}, error) {
+	replaced, err := applyFuncOnPath(inputState, path, expectedNode, replaceMapFieldValue, true)
```
[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/resolver/replace.go]: we have to ouput the minimal desiredState to to do the change, if we don't filter we will output all the interfaces from the machine, and that can be massive and useless.

File: nmpolicy/internal/resolver/resolver_test.go (hunk 2 above)
Comment (qinqon) [reply nmpolicy/internal/resolver/resolver_test.go:742]: The input currentState has no route with br1, where do I have to add it ?

File: nmpolicy/internal/parser/parser.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply nmpolicy/internal/parser/parser.go:236]: If we get `EOF` instead of the right hand string argument we don't return an error. We just procceed without it. Do I miss something?

File: nmpolicy/internal/resolver/replace.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlonaKaplan) [reply nmpolicy/internal/resolver/replace.go]: If the user wants to filter, he can do a filtering operation before the replace.
But now I think about it, since `replaceMapFieldValue` always returns a value it doesn't matter if we pass here `true` or `false`. The result would be the same. But I find it confusing to pass `true`. We are not filtering anything.

File: nmpolicy/internal/parser/parser.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/parser/parser.go:236]: I will add the missing replace failure unit tests and if needed fix it.

File: nmpolicy/internal/resolver/replace.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/resolver/replace.go]: btw this is old code, we have to pass a pair of flags `shouldFilterSlice bool, shouldFilterMap bool`

[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/resolver/replace.go]: apart from filtering they will change the value with the output form the function we pass they do both things.

[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/resolver/replace.go]: The code is like this 
```golang
for _, valueToCheck := range originalSlice {                                
        value, err := applyFuncOnPath(valueToCheck, path, expectedNode, funcToApply, false, false)
        if err != nil {                                                         
            return nil, err                                                     
        }                                                                       
        if value != nil {                                                       
            filteredSlice = append(filteredSlice, value)                        
        }                                                                       
    }                                                                           
                                                                                
    if len(filteredSlice) == 0 {                                                
        return nil, nil                                                         
    }                                                                           
                                                                                
    if shouldFilterSlice {                                                      
        return filteredSlice, nil                                               
    }                              
```
If we don't pass shouldFilterSlice it does not use the value returned by the function.

It feels very ad-hoc, we can generalize at follow up.

[the remark this answers is not in the document]
Comment (qinqon) [reply nmpolicy/internal/resolver/replace.go]: Got it, let's do only replace on replace, at the pipe PR we will tests it with the filter.

**Comment by qinqon** [comment]
/hold

Let's not filter on replace if users need it they will have to pipe it first.

File: nmpolicy/internal/resolver/filter.go [hunk 5]
```diff
@@ -23,7 +23,7 @@ import (
 )
 
 func filter(inputState map[string]interface{}, path ast.VariadicOperator, expectedNode ast.Node) (map[string]interface{}, error) {
-	filtered, err := applyFuncOnPath(inputState, path, expectedNode, mapContainsValue, true)
+	filtered, err := applyFuncOnMap(path, inputState, expectedNode, mapContainsValue, true, true)
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/filter.go:26]: I see `applyFuncOnMap` as an internal method that should be used only by `applyFuncOnPath`.
Anyway, how is this change related to the commit/PR?

File: nmpolicy/internal/resolver/path.go [hunk 6]
```diff
@@ -32,7 +32,7 @@ func applyFuncOnPath(inputState interface{},
 	path []ast.Node,
 	expectedNode ast.Node,
 	funcToApply func(map[string]interface{}, string, ast.Node) (interface{}, error),
-	shouldFilterSlice bool) (interface{}, error) {
+	shouldFilterSlice bool, shouldFilterMap bool) (interface{}, error) {
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/path.go:35]: IIUC, this change is for readability only? Right?
The old code also return all the matching slices as is (without any map/slice filtering). And all the maps till the matching slice were filtered.
If so, it is not related to this commit/PR. 

EDIT: unless you accept my comment about not filtering at all in case of a replace. Than I understand why the change is relevant here.

File: nmpolicy/internal/resolver/replace.go [hunk 7]
```diff
@@ -21,7 +21,7 @@ import (
 )
 
 func replace(inputState map[string]interface{}, path ast.VariadicOperator, expectedNode ast.Node) (map[string]interface{}, error) {
-	replaced, err := applyFuncOnPath(inputState, path, expectedNode, replaceMapFieldValue, true)
+	replaced, err := applyFuncOnMap(path, inputState, expectedNode, replaceMapFieldValue, true, true)
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/replace.go:24]: I think that in case of a replace we shouldn't filter anything. The output should similar to the input + the replacement.
When you do find and replace on a doc, you don't assume the docs will be filtered as well.

File: nmpolicy/internal/resolver/filter.go (hunk 5 above)
Comment (qinqon) [reply nmpolicy/internal/resolver/filter.go:26]: this is internal code, and we know that inputState is a map, does not make sense to call `applyFuncOnPath`, this way we remove one call.

File: nmpolicy/internal/resolver/path.go (hunk 6 above)
Comment (qinqon) [reply nmpolicy/internal/resolver/path.go:35]: I am chacing it to not filter on replace so we need it.

File: nmpolicy/internal/resolver/path.go [hunk 8]
```diff
@@ -57,32 +57,33 @@ func applyFuncOnSlice(originalSlice []interface{},
...
 		if err != nil {
 			return nil, err
 		}
-		if value != nil {
-			filteredSlice = append(filteredSlice, valueToCheck)
+		if valueAfterApply != nil {
+			sliceEmptyAfterApply = false
+			adjustedSlice = append(adjustedSlice, valueAfterApply)
+		} else if !shouldFilterSlice {
+			adjustedSlice = append(adjustedSlice, valueToCheck)
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/path.go:71]: missing `sliceEmptyAfterApply=false`.

File: nmpolicy/internal/resolver/path.go [hunk 9]
```diff
@@ -57,32 +57,33 @@ func applyFuncOnSlice(originalSlice []interface{},
...
 		}
 	}
 
-	if len(filteredSlice) == 0 {
+	if sliceEmptyAfterApply {
```
Comment (AlonaKaplan) [inline nmpolicy/internal/resolver/path.go:75]: I would keep it `len(adjustedSlice) == 0`.

File: nmpolicy/internal/resolver/path.go (hunk 8 above)
Comment (qinqon) [reply nmpolicy/internal/resolver/path.go:71]: Here valueAfterApply is nil we don't have to set the variable

File: nmpolicy/internal/resolver/path.go (hunk 9 above)
Comment (qinqon) [reply nmpolicy/internal/resolver/path.go:75]: Is no tthe same, adjustedSlice can be > 0 and sliceEmptyAfterApply false.

**Review by AlonaKaplan — APPROVED** [review]
